### PR TITLE
Use Url type in database structs (fixes #1364)

### LINKED ISSUE #1364: Directly use `Url` type in database structs
That would give us more type safety, as it ensures that we cant accidentally insert invalid Urls. It also means that we wont have to convert types manually all the time.

Problem is with these two errors:
`the trait AppearsOnTableschema::community::table is not implemented for Url`
`the trait diesel::Expression is not implemented for Url`

I didnt find any info in the docs about how to implement these traits. And its possible that we cant implement them in Lemmy, but they have to be implemented directly in diesel or url.

### PR DESCRIPTION
**Pull request by Nutomic**
As far as I can tell this implements all the required traits, but still fails with the following errors:

`the trait diesel::Queryable<diesel::sql_types::Text, Pg> is not implemented for (url::Url,)`

`the trait QueryFragment<Pg> is not implemented for Url`

[CI log with full errors](https://cloud.drone.io/LemmyNet/lemmy/300/1/6)

### CODE REVIEW

File: crates/db_schema/src/lib.rs [hunk 1]
```diff
@@ -1,11 +1,56 @@
...
+use serde::Serialize;
+use std::io::Write;
 
 pub mod schema;
 pub mod source;
 
+#[repr(transparent)]
+#[derive(Clone, PartialEq, Serialize, SqlType, Debug, AsExpression, QueryId, Queryable)]
+#[postgres(type_name = "Url")]
+pub struct Url(url::Url);
```
Comment (weiznich) [inline crates/db_schema/src/lib.rs:21]:
```suggestion
#[derive(Clone, PartialEq, Serialize, Debug, AsExpression, FromSqlRow)]
#[sql_type(type_name = "Text")]
pub struct Url(url::Url);
```

File: crates/db_schema/src/lib.rs [hunk 2]
```diff
@@ -1,11 +1,56 @@
...
+    let str = String::from_sql(bytes)?;
+    Ok(Url(url::Url::parse(&str)?))
+  }
+}
+
+impl Expression for Url {
+  type SqlType = Text;
+}
+
+impl<QS> AppearsOnTable<QS> for Url {}
```
Comment (weiznich) [inline crates/db_schema/src/lib.rs:46]:
```suggestion

```

File: crates/db_schema/src/lib.rs (hunk 1 above)
Comment (Nutomic) [reply crates/db_schema/src/lib.rs:21]: Perfect, thank you!

**Comment by Nutomic** [comment]
Okay I converted actor_id and ap_id columns. Tests are failing, not really sure why. There are also other fields which should be converted, eg `Post::url` or `Post::thumbnail_url`.

**Comment by Nutomic** [comment]
Integration tests fixed, but most of the federation tests are still failing.

**Comment by Nutomic** [comment]
One federation test failed on x86, but everything passed on ARM. After restarting the tests, everything passed so this seems good to merge.

**Comment by dessalines** [comment]
I gonna re-run the tests one more time to make sure.

File: crates/db_schema/src/lib.rs [hunk 3]
```diff
@@ -2,10 +2,69 @@
...
+    url.0
+  }
+}
+
+impl From<url::Url> for Url {
+  fn from(url: url::Url) -> Self {
+    Url(url)
+  }
+}
+
```
Comment (dessalines) [inline crates/db_schema/src/lib.rs:67]: Looks good.

**Review by dessalines — APPROVED** [review]

**Pull request merged**